### PR TITLE
feat: Support TS satisfies expression

### LINKED ISSUE #3748: 🐛 Update parser to support `satisfies` keyword (Typescript 4.9+)
### Environment information

```block
CLI:
  Version:              10.0.1
  Color support:        true

Platform:
  CPU Architecture:     x86_64
  OS:                   windows

Environment:
  ROME_LOG_DIR:         unset
  NO_COLOR:             unset
  TERM:                 "cygwin"

Rome Configuration:
  Status:               loaded
  Formatter disabled:   false
  Linter disabled:      true

Workspace:
  Open Documents:       0

Discovering running Rome servers...

Server:
  Status:               stopped
```

### What happened?

Today ([Nov 15 2022](https://github.com/microsoft/TypeScript/issues/50457)) Typescript 4.9 reaches planned release date. The brand new [`satisfies`](https://devblogs.microsoft.com/typescript/announcing-typescript-4-9-beta/#the-satisfies-operator) keyword is coming into play.

The updates for that keyword has already been included before in other tools like esbuild (since 0.5.14) and SWC (since 1.3.15).

Rome tools right now fail [(playground link)](https://docs.rome.tools/playground/?code=dAB5AHAAZQAgAEMAbwBsAG8AcgBzACAAPQAgACIAcgBlAGQAIgAgAHwAIAAiAGcAcgBlAGUAbgAiACAAfAAgACIAYgBsAHUAZQAiADsACgAKAHQAeQBwAGUAIABSAEcAQgAgAD0AIABbAHIAZQBkADoAIABuAHUAbQBiAGUAcgAsACAAZwByAGUAZQBuADoAIABuAHUAbQBiAGUAcgAsACAAYgBsAHUAZQA6ACAAbgB1AG0AYgBlAHIAXQA7AAoACgBjAG8AbgBzAHQAIABwAGEAbABlAHQAdABlACAAPQAgAHsACgAgACAAIAAgAHIAZQBkADoAIABbADIANQA1ACwAIAAwACwAIAAwAF0ALAAKACAAIAAgACAAZwByAGUAZQBuADoAIAAiACMAMAAwAGYAZgAwADAAIgAsAAoAIAAgACAAIABiAGwAZQB1ADoAIABbADAALAAgADAALAAgADIANQA1AF0ACgAvAC8AIAAgAH4AfgB%2BAH4AIABUAGgAZQAgAHQAeQBwAG8AIABpAHMAIABuAG8AdwAgAGMAYQB1AGcAaAB0ACEACgB9ACAAcwBhAHQAaQBzAGYAaQBlAHMAIABSAGUAYwBvAHIAZAA8AEMAbwBsAG8AcgBzACwAIABzAHQAcgBpAG4AZwAgAHwAIABSAEcAQgA%2BADsA&typescript=true&jsx=false) to parse code containing such keyword:

<details>
  <summary>Error log</summary>

```
> rome format ./src --write

./index.ts:12:3 parse ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

  × Expected a semicolon or an implicit semicolon after a statement, but found none

    10 │   blue: [0, 0, 255],
    11 │ //  ~~~~ The typo is now caught!
  > 12 │ } satisfies Record<Colors, string | RGB>;
       │   ^^^^^^^^^
    13 │ 

  i An explicit or implicit semicolon is expected here...

    10 │   blue: [0, 0, 255],
    11 │ //  ~~~~ The typo is now caught!
  > 12 │ } satisfies Record<Colors, string | RGB>;
       │   ^^^^^^^^^
    13 │ 

  i ...Which is required to end this statement

     5 │ type RGB = [red: number, green: number, blue: number];
     6 │ 
   > 7 │ const palette = {
       │ ^^^^^^^^^^^^^^^^^
   > 8 │   red: [255, 0, 0],
   > 9 │   green: '#00ff00',
  > 10 │   blue: [0, 0, 255],
  > 11 │ //  ~~~~ The typo is now caught!
  > 12 │ } satisfies Record<Colors, string | RGB>;
       │ ^^
    13 │ 


```

</details>

### Expected result

`satisfies` keyword is expected to be parsed correctly.

_I guess it should not be very complex, but unfortunately I'm not familiar enough with Rust. Furthermore, it is code related to the parser core (and files [generated](https://github.com/rome/tools/blob/main/crates/rome_js_syntax/src/generated/kind.rs#L607) from I have no idea what source).
Therefore I won't try fixing it and filing a PR myself_ 

### Code of Conduct

- [X] I agree to follow Rome's Code of Conduct

### PR DESCRIPTION
**Pull request by nissy-dev**
Fix #3748

## Summary

This PR is initial implementation for satisfies expression released in TS.4.9.
This is almost same as the implementation of TsAsExpression.

## Test Plan

I added some parser  and formatter tests. Also, I updated prettier tests about satisfies expression.

I was able to pass most of the Prettier test cases except for `comments-unstable.ts`. As for `comments-unstable.ts`, I don't understand how to pass it.  I would appreciate it if you could give me some advice.

### CODE REVIEW

**Review by MichaReiser — APPROVED** [review]
This is awesome work! Well done

**Comment by MichaReiser** [comment]
There's one failing test that you need to update before we can merge this.

File: crates/rome_js_formatter/src/generated.rs [hunk 1]
```diff
@@ -2756,6 +2756,44 @@ impl IntoFormat<crate::JsFormatContext> for rome_js_syntax::TsAsExpression {
         )
     }
 }
+impl FormatRule<rome_js_syntax::TsSatisfiesExpression>
```
Comment (ematipico) [inline crates/rome_js_formatter/src/generated.rs:2759]: @MichaReiser should we add this file to `.gitattributes`?

File: crates/rome_js_parser/src/syntax/expr.rs [hunk 2]
```diff
@@ -598,6 +599,32 @@ fn parse_binary_or_logical_expression_recursive(
...
+        // satisfies;
+        // let precedence = "hello" satisfies string + 3 satisfies number satisfies number;
+        if op == T![satisfies] {
+            parse_ts_type(p).or_add_diagnostic(p, expected_ts_type);
+            let mut satisfies_expression = m.complete(p, TS_SATISFIES_EXPRESSION);
+
+            if TypeScript.is_unsupported(p) {
+                p.error(ts_only_syntax_error(
+                    p,
+                    "'satisfies' expression",
```
Comment (ematipico) [inline crates/rome_js_parser/src/syntax/expr.rs:619]: Can we add a test where we emit this new diagnostic? Ideally, just a JS file with this new syntax, should be enough.

**Review by ematipico — APPROVED** [review]

File: crates/rome_js_parser/src/syntax/expr.rs (hunk 2 above)
Comment (nissy-dev) [reply crates/rome_js_parser/src/syntax/expr.rs:622]: I added in 8e31e54 !

**Comment by nissy-dev** [comment]
> There's one failing test that needs updating before merging the PR

After checking one failing test, I confirmed the tests will pass by commenting out [this line](https://github.com/rome/tools/blob/67a888158ace04201f5c1d97c5b8898bd89405bf/crates/rome_js_formatter/tests/check_reformat.rs#L71) in check_reformat.rs. And then, I confirmed the formatted result is not different from prettiter's one so I removed `comments-unstable.ts.snap`.  Currently, unrelated tests failed.

**Comment by MichaReiser** [comment]
> > There's one failing test that needs updating before merging the PR
> 
> After checking one failing test, I confirmed the tests will pass by commenting out [this line](https://github.com/rome/tools/blob/67a888158ace04201f5c1d97c5b8898bd89405bf/crates/rome_js_formatter/tests/check_reformat.rs#L71) in check_reformat.rs. And then, I confirmed the formatted result is not different from prettiter's one so I removed `comments-unstable.ts.snap`. Currently, only unrelated tests failed.

I pushed a commit that handles leading own line comments by creating an indent around the type. The resulting formatting differs from Prettier but is stable.

**Pull request merged**